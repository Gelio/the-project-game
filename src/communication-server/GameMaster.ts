import { LoggerInstance } from 'winston';

import { Communicator } from '../common/Communicator';
import { CustomEventEmitter } from '../common/CustomEventEmitter';
import { COMMUNICATION_SERVER_ID, GAME_MASTER_ID } from '../common/EntityIds';
import { REQUEST_TYPE } from '../common/REQUEST_TYPE';

import { stringifySchemaValidationErrors } from '../common/logging/stringifySchemaValidationErrors';

import { Message } from '../interfaces/Message';
import { MessageWithRecipient } from '../interfaces/MessageWithRecipient';
import { UnregisterGameResponse } from '../interfaces/responses/UnregisterGameResponse';

import { Game } from './Game';
import { MessageRouter } from './MessageRouter';
import { SimpleMessageValidator } from './SimpleMessageValidator';

export class GameMaster extends CustomEventEmitter {
  public readonly game: Game;
  public readonly communicator: Communicator;

  private readonly messageRouter: MessageRouter;
  private readonly logger: LoggerInstance;
  private readonly messageValidator: SimpleMessageValidator;

  constructor(
    communicator: Communicator,
    messageRouter: MessageRouter,
    logger: LoggerInstance,
    game: Game,
    messageValidator: SimpleMessageValidator
  ) {
    super();

    this.communicator = communicator;
    this.messageRouter = messageRouter;
    this.logger = logger;
    this.game = game;
    this.messageValidator = messageValidator;

    this.handleMessage = this.handleMessage.bind(this);
    this.onDisconnected = this.onDisconnected.bind(this);
  }

  public init() {
    this.messageRouter.registerGameMasterCommunicator(
      this.game.gameDefinition.name,
      this.communicator
    );
    this.communicator.on('message', this.handleMessage);
    this.communicator.once('destroy', this.onDisconnected);
  }

  public onDisconnected() {
    this.unbindListeners();

    const gameName = this.game.gameDefinition.name;
    this.logger.info(`Game Master from game ${gameName} disconnected`);

    this.emit('disconnect');
    this.game.destroy();
  }

  public onGameFinished() {
    this.unbindListeners();

    const gameName = this.game.gameDefinition.name;
    this.logger.info(`Game ${gameName} finished`);

    this.emit('gameFinish');
    this.game.finish();
  }

  public sendMessage(message: Message<any>) {
    this.communicator.sendMessage(message);
  }

  private handleMessage(message: MessageWithRecipient<any>) {
    if (!this.messageValidator(message)) {
      const gameName = this.game.gameDefinition.name;
      this.logger.warn(`Invalid message received from GM handling game "${gameName}"`);

      this.logger.verbose('Message:');
      this.logger.verbose(JSON.stringify(message));

      const stringifiedErrors = stringifySchemaValidationErrors(this.messageValidator.errors || []);
      this.logger.verbose(stringifiedErrors);

      return;
    }

    if (message.type === REQUEST_TYPE.UNREGISTER_GAME_REQUEST) {
      this.onGameFinished();

      const response: UnregisterGameResponse = {
        senderId: COMMUNICATION_SERVER_ID,
        recipientId: GAME_MASTER_ID,
        type: 'UNREGISTER_GAME_RESPONSE',
        payload: {
          unregistered: true
        }
      };
      this.sendMessage(response);

      return;
    }

    this.messageRouter.sendMessageToPlayer(message);
  }

  private unbindListeners() {
    this.communicator.removeListener('destroy', this.onDisconnected);
    this.communicator.removeListener('message', this.handleMessage);

    this.messageRouter.unregisterGameMasterCommunicator(this.game.gameDefinition.name);
  }
}
