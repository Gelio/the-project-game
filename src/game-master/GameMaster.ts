import { createConnection } from 'net';
import { LoggerInstance } from 'winston';

import { bindObjectProperties } from '../common/bindObjectProperties';
import { Communicator } from '../common/Communicator';
import { LoggerFactory } from '../common/logging/LoggerFactory';
import { UITransport } from '../common/logging/UITransport';

import { ActionDelays } from '../interfaces/ActionDelays';
import { BoardSize } from '../interfaces/BoardSize';
import { GameDefinition, TeamSizes } from '../interfaces/GameDefinition';
import { Message } from '../interfaces/Message';
import { Service } from '../interfaces/Service';

import { ActionInvalidMessage } from '../interfaces/messages/ActionInvalidMessage';
import { ActionValidMessage } from '../interfaces/messages/ActionValidMessage';
import {
  GameStartedMessage,
  GameStartedMessagePayload
} from '../interfaces/messages/GameStartedMessage';
import { PlayerAcceptedMessage } from '../interfaces/messages/PlayerAcceptedMessage';
import { PlayerDisconnectedMessage } from '../interfaces/messages/PlayerDisconnectedMessage';
import { PlayerHelloMessage } from '../interfaces/messages/PlayerHelloMessage';
import { PlayerRejectedMessage } from '../interfaces/messages/PlayerRejectedMessage';
import { MessageWithRecipient } from '../interfaces/MessageWithRecipient';
import { RegisterGameRequest } from '../interfaces/requests/RegisterGameRequest';
import { RegisterGameResponse } from '../interfaces/responses/RegisterGameResponse';

import { registerUncaughtExceptionHandler } from '../registerUncaughtExceptionHandler';

import { PeriodicPieceGenerator } from './board-generation/PeriodicPieceGenerator';

import { Game } from './Game';
import { GameMasterState } from './GameMasterState';
import { Player } from './Player';
import { PlayersContainer } from './PlayersContainer';

import { UIController } from './ui/UIController';

export interface GameMasterOptions {
  serverHostname: string;
  serverPort: number;
  gameName: string;
  gameDescription: string;
  gamesLimit: number;
  teamSizes: TeamSizes;
  pointsLimit: number;
  boardSize: BoardSize;
  shamChance: number;
  generatePiecesInterval: number;
  piecesLimit: number;
  resultFileName: string;
  actionDelays: ActionDelays;
  timeout: number;
  registrationTriesLimit: number;
}

export class GameMaster implements Service {
  private readonly options: GameMasterOptions;
  private communicator: Communicator;
  private game: Game;
  private playersContainer: PlayersContainer;
  private state: GameMasterState;

  private readonly uiController: UIController;
  private readonly loggerFactory: LoggerFactory;
  private periodicPieceGenerator: PeriodicPieceGenerator;
  private logger: LoggerInstance;

  private readonly messageHandlers: { [type: string]: Function } = {
    PLAYER_HELLO: this.handlePlayerHelloMessage,
    PLAYER_DISCONNECTED: this.handlePlayerDisconnectedMessage
  };

  constructor(
    options: GameMasterOptions,
    uiController: UIController,
    loggerFactory: LoggerFactory
  ) {
    this.options = options;
    this.uiController = uiController;
    this.loggerFactory = loggerFactory;
    this.state = GameMasterState.Connecting;

    bindObjectProperties(this.messageHandlers, this);
    this.destroy = this.destroy.bind(this);
    this.handleMessage = this.handleMessage.bind(this);
  }

  public init() {
    this.initUI();
    this.initLogger();
    this.initGame();

    const { serverHostname, serverPort } = this.options;

    const socket = createConnection(
      {
        host: serverHostname,
        port: serverPort
      },
      () => {
        this.logger.info(`Connected to the server at ${serverHostname}:${serverPort}`);
      }
    );

    this.communicator = new Communicator(socket, this.logger);
    this.communicator.bindListeners();

    this.communicator.once('close', this.handleServerDisconnection.bind(this));

    this.registerGame();
  }

  public destroy() {
    if (this.game.hasStarted) {
      this.stopGame();
    }

    this.communicator.destroy();
    this.uiController.destroy();
  }

  private handleServerDisconnection() {
    this.logger.warn('Disconnected from the server. Closing...');

    this.destroy();
  }

  private async handleMessage<T>(message: Message<T>) {
    if (this.messageHandlers[message.type] !== undefined) {
      return this.messageHandlers[message.type](message);
    }

    const result = this.game.processMessage(message);
    if (!result.valid) {
      const actionInvalidMessage: ActionInvalidMessage = {
        type: 'ACTION_INVALID',
        recipientId: message.senderId,
        senderId: -1,
        payload: {
          reason: result.reason
        }
      };

      return this.communicator.sendMessage(actionInvalidMessage);
    }

    const actionValidMessage: ActionValidMessage = {
      type: 'ACTION_VALID',
      recipientId: message.senderId,
      senderId: -1,
      payload: {
        delay: result.delay
      }
    };

    this.communicator.sendMessage(actionValidMessage);
    this.communicator.sendMessage(await result.responseMessage);
  }

  private handlePlayerHelloMessage(message: PlayerHelloMessage) {
    this.logger.verbose(`Received player ${message.payload.temporaryId} hello message`);

    try {
      const assignedPlayerId = this.tryAcceptPlayer(message);

      const playerAcceptedMessage: PlayerAcceptedMessage = {
        type: 'PLAYER_ACCEPTED',
        senderId: -1,
        recipientId: message.payload.temporaryId,
        payload: {
          assignedPlayerId
        }
      };

      this.communicator.sendMessage(playerAcceptedMessage);
      this.tryStartGame();
    } catch (e) {
      const error: Error = e;

      this.logger.verbose(`Player ${message.payload.temporaryId} rejected. Reason: ${e.message}`);

      const playerRejectedMessage: PlayerRejectedMessage = {
        type: 'PLAYER_REJECTED',
        senderId: -1,
        recipientId: message.payload.temporaryId,
        payload: {
          reason: error.message
        }
      };

      return this.communicator.sendMessage(playerRejectedMessage);
    }
  }

  private async registerGame() {
    const game: GameDefinition = {
      name: this.options.gameName,
      description: this.options.gameDescription,
      teamSizes: this.options.teamSizes,
      boardSize: this.options.boardSize,
      goalLimit: this.options.pointsLimit,
      delays: this.options.actionDelays
    };
    const registerGameMessage: RegisterGameRequest = {
      type: 'REGISTER_GAME_REQUEST',
      senderId: -1,
      payload: game
    };

    this.communicator.sendMessage(registerGameMessage);

    const listGamesResponse = <RegisterGameResponse>await this.communicator.waitForSpecificMessage(
      (msg: MessageWithRecipient<RegisterGameResponse>) => msg.type === 'REGISTER_GAME_RESPONSE'
    );

    this.handleRegisterGameResponse(listGamesResponse);
  }

  private handleRegisterGameResponse(message: RegisterGameResponse) {
    if (message.payload.registered === true) {
      this.logger.verbose('Received game registered response');
      this.updateState(GameMasterState.WaitingForPlayers);

      this.communicator.on('message', this.handleMessage);

      return;
    }

    if (this.options.registrationTriesLimit === 0) {
      throw new Error('Failed to register new game, limit of tries reached');
    }

    this.options.registrationTriesLimit--;

    this.logger.error(
      `Failed to register new game! Next attempt will be made in 10 seconds. Attempts left: ${
        this.options.registrationTriesLimit
      }`
    );

    setTimeout(() => this.registerGame(), 10000);
  }

  private tryAcceptPlayer(message: PlayerHelloMessage) {
    const teamPlayers = this.game.playersContainer.getPlayersFromTeam(message.payload.teamId);

    if (this.game.hasStarted) {
      const disconnectedPlayer = teamPlayers.find(
        player => !player.isConnected && player.isLeader === message.payload.isLeader
      );

      if (!disconnectedPlayer) {
        throw new Error('Game already started and no more slots free');
      }

      disconnectedPlayer.isConnected = true;

      return disconnectedPlayer.playerId;
    }

    if (teamPlayers.length >= this.options.teamSizes[message.payload.teamId]) {
      throw new Error('Team is full');
    }

    if (message.payload.isLeader && teamPlayers.find(player => player.isLeader)) {
      throw new Error('Team already has a leader');
    }

    if (
      !message.payload.isLeader &&
      teamPlayers.length + 1 === this.options.teamSizes[message.payload.teamId]
    ) {
      throw new Error('Team does not have a leader');
    }

    const newPlayer = new Player();
    newPlayer.playerId = this.game.getNextPlayerId();
    newPlayer.teamId = message.payload.teamId;
    newPlayer.isLeader = message.payload.isLeader;
    newPlayer.isBusy = false;
    newPlayer.isConnected = true;

    this.game.addPlayer(newPlayer);

    return newPlayer.playerId;
  }

  private handlePlayerDisconnectedMessage(message: PlayerDisconnectedMessage) {
    this.logger.verbose('Received player disconnected message');
    const disconnectedPlayer = this.playersContainer.getPlayerById(message.payload.playerId);

    if (!this.game.hasStarted) {
      if (disconnectedPlayer) {
        this.game.removePlayer(disconnectedPlayer);
        this.uiController.updateBoard(this.game.board);
      }

      return;
    }

    if (!disconnectedPlayer) {
      return;
    }

    disconnectedPlayer.isConnected = false;

    const connectedPlayers = this.game.playersContainer.getConnectedPlayers();
    if (connectedPlayers.length === 0) {
      this.logger.info('All players disconnected, disconnecting from the server');
      this.destroy();
      //TODO check if GM should try to start new game
    }
  }

  private initGame() {
    this.playersContainer = new PlayersContainer();

    this.game = new Game(
      this.options.boardSize,
      this.options.pointsLimit,
      this.logger,
      this.uiController,
      this.playersContainer
    );
    this.uiController.updateBoard(this.game.board);

    this.periodicPieceGenerator = new PeriodicPieceGenerator(
      this.game,
      {
        checkInterval: this.options.generatePiecesInterval,
        piecesLimit: this.options.piecesLimit,
        shamChance: this.options.shamChance
      },
      this.logger
    );
  }

  private tryStartGame() {
    const connectedPlayersCount = this.playersContainer.players.length;
    const requiredPlayersCount = this.options.teamSizes['1'] + this.options.teamSizes['2'];

    if (connectedPlayersCount < requiredPlayersCount) {
      return;
    }

    try {
      this.startGame();
    } catch (error) {
      this.logger.error(`Cannot start game - ${error.message}`);
    }
  }

  private startGame() {
    this.logger.info('Game is starting...');

    const team1Players = this.game.playersContainer.getPlayersFromTeam(1);
    const team2Players = this.game.playersContainer.getPlayersFromTeam(2);
    const team1Leader = team1Players.find(player => player.isLeader);
    const team2Leader = team2Players.find(player => player.isLeader);

    if (!team1Leader || !team2Leader) {
      throw new Error('Game cannot start without both leaders');
    }

    this.game.setPlayersPositions();
    this.periodicPieceGenerator.init();
    const gameStartedPayload: GameStartedMessagePayload = {
      teamInfo: {
        1: {
          players: team1Players.map(player => player.playerId),
          leaderId: team1Leader.playerId
        },
        2: {
          players: team2Players.map(player => player.playerId),
          leaderId: team2Leader.playerId
        }
      }
    };
    this.game.start();
    this.updateState(GameMasterState.InGame);

    this.playersContainer.players.forEach(player => {
      const message: GameStartedMessage = {
        senderId: -1,
        recipientId: player.playerId,
        type: 'GAME_STARTED',
        payload: gameStartedPayload
      };

      this.communicator.sendMessage(message);
    });

    this.logger.info('Game started');
  }

  private stopGame() {
    this.periodicPieceGenerator.destroy();
    this.game.stop();
    this.updateState(GameMasterState.Finished);
  }

  private updateState(state: GameMasterState) {
    this.state = state;
    this.uiController.render();
  }

  private initUI() {
    this.uiController.init();
    this.updateState(this.state);
  }

  private initLogger() {
    if (!this.uiController) {
      throw new Error('Cannot create logger without UI controller');
    }

    const uiTransport = new UITransport(this.uiController.log.bind(this.uiController));
    this.logger = this.loggerFactory.createLogger([uiTransport]);
    registerUncaughtExceptionHandler(this.logger);
    this.logger.info('Logger initiated');
  }
}
