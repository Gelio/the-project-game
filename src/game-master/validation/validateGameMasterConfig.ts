import { GameMasterOptions } from '../GameMaster';

import { validateJSON } from '../../common/validation/validateJSON';

// tslint:disable-next-line no-require-imports no-var-requires
const gmConfigSchema = require('../../../schemas/configs/game-master-config.json');

export async function validateGameMasterConfig(config: GameMasterOptions) {
  await validateJSON(gmConfigSchema, config);

  const taskAreaSize = config.boardSize.taskArea * config.boardSize.x;
  const goalAreaSize = config.boardSize.goalArea * config.boardSize.x;

  if (config.piecesLimit > taskAreaSize) {
    throw new Error(
      'There cannot be more pieces on the board than there are tiles in the task area'
    );
  }

  if (config.teamSizes[1] > goalAreaSize) {
    throw new Error('Team 1 has a greater capacity than there are tiles in its area');
  }

  if (config.teamSizes[2] > goalAreaSize) {
    throw new Error('Team 2 has a greater capacity than there are tiles in its area');
  }
}
