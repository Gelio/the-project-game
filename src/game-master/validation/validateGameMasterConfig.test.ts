import { GameMasterOptions } from '../GameMaster';

import { validateGameMasterConfig } from './validateGameMasterConfig';

function getMockGameMasterConfig(): GameMasterOptions {
  return {
    serverHostname: 'localhost',
    serverPort: 4200,
    gameName: 'the project game',
    gameDescription: 'High sham chance, many pieces',
    gamesLimit: 2,
    teamSizes: {
      1: 1,
      2: 1
    },
    pointsLimit: 1,
    boardSize: {
      x: 4,
      taskArea: 3,
      goalArea: 2
    },
    shamChance: 0.5,
    generatePiecesInterval: 1500,
    piecesLimit: 5,
    logsDirectory: 'logs',
    actionDelays: {
      move: 35,
      pick: 200,
      discover: 50,
      destroy: 300,
      test: 400,
      communicationRequest: 1000,
      communicationAccept: 1000,
      place: 200
    },
    timeout: 10000,
    registrationTriesLimit: 5,
    registerGameInterval: 10000
  };
}

describe('[GM] validateGameMasterConfig', () => {
  it('should validate config against a schema', () => {
    const config = getMockGameMasterConfig();

    config.serverPort = 50000000000;

    return expect(validateGameMasterConfig(config)).rejects.toMatchSnapshot();
  });

  it('should throw an error when pieces limit is too large', () => {
    const config = getMockGameMasterConfig();

    config.piecesLimit = 13245645;

    return expect(validateGameMasterConfig(config)).rejects.toMatchSnapshot();
  });

  it('should throw an error when team 1 size is too large', () => {
    const config = getMockGameMasterConfig();

    config.teamSizes[1] = 13245645;

    return expect(validateGameMasterConfig(config)).rejects.toMatchSnapshot();
  });

  it('should throw an error when team 2 size is too large', () => {
    const config = getMockGameMasterConfig();

    config.teamSizes[2] = 13245645;

    return expect(validateGameMasterConfig(config)).rejects.toMatchSnapshot();
  });

  it('should not throw for a valid config', async done => {
    const config = getMockGameMasterConfig();

    try {
      await validateGameMasterConfig(config);
      done();
    } catch (error) {
      done.fail('Validation failed');
    }
  });
});
