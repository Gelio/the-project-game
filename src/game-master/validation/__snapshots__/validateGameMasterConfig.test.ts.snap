// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[GM] validateGameMasterConfig should throw an error when pieces limit is too large 1`] = `[Error: There cannot be more pieces on the board than there are tiles in the task area]`;

exports[`[GM] validateGameMasterConfig should throw an error when team 1 size is too large 1`] = `[Error: Team 1 has a greater capacity than there are tiles in its area]`;

exports[`[GM] validateGameMasterConfig should throw an error when team 2 size is too large 1`] = `[Error: Team 2 has a greater capacity than there are tiles in its area]`;

exports[`[GM] validateGameMasterConfig should validate config against a schema 1`] = `
Array [
  Object {
    "dataPath": ".serverPort",
    "keyword": "maximum",
    "message": "should be <= 49151",
    "params": Object {
      "comparison": "<=",
      "exclusive": false,
      "limit": 49151,
    },
    "schemaPath": "../common/common-defs.json#/definitions/port/maximum",
  },
]
`;
